cmd/clusterctl/client/cluster/topology_plan_test.go

```diff
@@ -0,0 +1,492 @@
+	//go:embed assets/topology-test/modified-CP-dockermachinetemplate.yaml
+	modifiedDockerMachineTemplateYAML []byte
+
+	//go:embed assets/topology-test/objects-in-different-namespaces.yaml
+	objsInDifferentNamespacesYAML []byte
+)
+
+func Test_topologyClient_Plan(t *testing.T) {
+	type args struct {
+		in *TopologyPlanInput
+	}
+	type item struct {
+		kind       string
+		namespace  string
+		namePrefix string
+	}
+	type out struct {
+		affectedClusters       []client.ObjectKey
+		affectedClusterClasses []client.ObjectKey
+		reconciledCluster      *client.ObjectKey
+		created                []item
+		modified               []item
+		deleted                []item
+	}
+	tests := []struct {
+		name            string
+		existingObjects []*unstructured.Unstructured
+		args            args
+		want            out
+		wantErr         bool
+	}{
+		{
+			name: "Input with new ClusterClass and new Cluster",
+			args: args{
+				in: func() *TopologyPlanInput {
+					objs, err := utilyaml.ToUnstructured(newClusterClassAndClusterYAML)
+					if err != nil {
+						panic(err)
+					}
+					return &TopologyPlanInput{
+						Objs: convertToPtrSlice(objs),
+					}
+				}(),
+			},
+			want: out{
+				created: []item{
+					{kind: "DockerCluster", namespace: "default", namePrefix: "my-cluster-"},
+					{kind: "DockerMachineTemplate", namespace: "default", namePrefix: "my-cluster-md-0-"},
+					{kind: "DockerMachineTemplate", namespace: "default", namePrefix: "my-cluster-md-1-"},
+					{kind: "DockerMachineTemplate", namespace: "default", namePrefix: "my-cluster-control-plane-"},
+					{kind: "KubeadmConfigTemplate", namespace: "default", namePrefix: "my-cluster-md-0-bootstrap-"},
+					{kind: "KubeadmConfigTemplate", namespace: "default", namePrefix: "my-cluster-md-1-bootstrap-"},
+					{kind: "KubeadmControlPlane", namespace: "default", namePrefix: "my-cluster-"},
+					{kind: "MachineDeployment", namespace: "default", namePrefix: "my-cluster-md-0-"},
+					{kind: "MachineDeployment", namespace: "default", namePrefix: "my-cluster-md-1-"},
+				},
+				modified: []item{
+					{kind: "Cluster", namespace: "default", namePrefix: "my-cluster"},
+				},
+				affectedClusters: func() []client.ObjectKey {
+					cluster := client.ObjectKey{Namespace: "default", Name: "my-cluster"}
+					return []client.ObjectKey{cluster}
+				}(),
+				affectedClusterClasses: func() []client.ObjectKey {
+					cc := client.ObjectKey{Namespace: "default", Name: "my-cluster-class"}
+					return []client.ObjectKey{cc}
+				}(),
+				reconciledCluster: &client.ObjectKey{Namespace: "default", Name: "my-cluster"},
+			},
+			wantErr: false,
+		},
+		{
+			name: "Modifying an existing Cluster",
+			existingObjects: func() []*unstructured.Unstructured {
+				objects := []unstructured.Unstructured{}
+
+				// Add CRDs
+				objs, err := utilyaml.ToUnstructured(mockCRDsYAML)
+				if err != nil {
+					panic(err)
+				}
```

Comment:
Nit, Should we make this a `mustToUnstructured` util func?

So then you can just write:
```go
objects = append(objects, mustToUnstructured(mockCRDsYAML)...)
```

Reply:
Additionally you can add another util func on top of that which takes a vararg of YAMLs and returns `[]*unstructured.Unstructured`, i.e. it would do a bunch of append mustToUnstructured + convertToPtrSlice

I think this would help avoid clutter and repetition in the test table

Reply:
Ideally you can shrink it to:
```go
existingObjects: mustYAMLToUnstructured(mockCRDsYAML, existingMyClusterClassYAML, existingMyClusterYAML)
```

Reply:
good suggestion. +1.